### ISSUE TITLE
Evaluating functions with partial arguments

### ISSUE THREAD

**Issue by ttuegel** [runtimeverification/haskell-backend#1978]
After #1911 

A common difficulty is evaluating functions with partial arguments. Consider a function pattern,

```
σ(φ₁, ...)
```

where `φᵢ` are not known to be defined. The rules defining `σ` will generally be written with element variables in the argument positions, requiring that the arguments be defined to apply the rule.

After implementing #1911, we can use this simple identity to always evaluate functions with partial arguments:

```
φ = φ ∧ ⌈φ⌉
```

We apply this identity to the arguments:

```
σ(φ₁, ...) = σ(φ₁ ∧ ⌈φ₁⌉, ...) = σ(φ₁, ...) ∧ ⌈φ₁⌉ ∧ ...
```

At this point, we may infer that the `φᵢ` in argument positions are defined, allowing us to apply most function rules even if the arguments are partial.

**Comment by ttuegel** [comment]
In [`makeExpandedApplication`](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Simplification/Application.hs#L145) we should use `mkDefined` and add side conditions for each argument saying it is defined.

**Comment by ttuegel** [comment]
I want to continue to put this off until after #2196.

**Comment by ttuegel** [comment]
After #2380, I want to do several different things.

1. Move `Kore.Step.Function` to `Kore.Function`. This is a significant domain in symbolic execution, and it should be recognized as such.
2. Likewise, come up with a better name for `Kore.Step.Axiom`. This doesn't even cover all axioms, only equations! And, it covers built-in simplification routines, not only user-defined equations.
1. In `Kore.Equation.Application`, separate matching/unification from checking requirements.
2. For `Kore.Step.Axiom.EvaluationStrategy.definitionEvaluation`, use unification to apply equations, but make any `\ceil` conditions part of the `ensures`, not the `requires`. Other conditions generated from unification remain in `requires`.
3. Separate `Kore.Step.Axiom.EvaluationStrategy.simplificationEvaluation` into cases with symbols at the top and cases with `\ceil` at the top. Both cases use matching to apply equations, but when symbols are at the top, make `\ceil` conditions part of the ensures (like `definitionEvaluation`); otherwise, the `\ceil` conditions remain in the `requires`.

In short, if `f(_) = R`, the backend will currently only evaluate `f(φ)` if it can infer `\ceil(φ)`. Actually, we can evaluate `f(φ)` to `R \and \ceil(φ)`, which is what the user wanted anyway.

**Comment by ana-pantilie** [comment]
@ttuegel suggests that we add definedness conditions of the `term` [here](https://github.com/kframework/kore/blob/3173ec42e2a342e52c5f712d1ff4b6c64b2eb107/kore/src/Kore/Step/Simplification/Application.hs#L127) to the `predicate` and the `SideCondition` (in the `definedTerms` field). We need to see how this will affect performance.

**Comment by ana-pantilie** [comment]
Do we still need this?
